packages/minimongo/upsert_document.js

```diff
@@ -0,0 +1,123 @@
+// Creating a document from an upsert is quite tricky.
+// E.g. this selector: {"$or": [{"b.foo": {"$all": ["bar"]}}]}, should result in: {"b.foo": "bar"}
+// But this selector: {"$or": [{"b": {"foo": {"$all": ["bar"]}}}]} should throw an error
+
+// Some rules (found mainly with trial & error, so there might be more):
+// - handle all childs of $and (or implicit $and)
+// - handle $or nodes with exactly 1 child
+// - ignore $or nodes with more than 1 child
+// - ignore $nor and $not nodes
+// - throw when a value can not be set unambiguously
+// - every value for $all should be dealt with as separate $eq-s
+// - threat all children of $all as $eq setters (=> set if $all.length === 1, otherwise throw error)
+// - you can not mix '$'-prefixed keys and non-'$'-prefixed keys
+// - you can only have dotted keys on a root-level
+// - you can not have '$'-prefixed keys more than one-level deep in an object
+
+// Fills a document with certain fields from an upsert selector
+populateDocumentWithQueryFields = function (query, document = {}) {
+
+    if (Object.getPrototypeOf(query) === Object.prototype) {
+        // handle implicit $and
+        for (let [key, value] of Object.entries(query)) {
```

Comment:
Using `let [key, value] ...` here works, but using `Object.keys(query).forEach(...` is preferred (this was mentioned in another [code review](https://github.com/meteor/meteor/pull/8643#pullrequestreview-36055174), which I discussed with @benjamn afterwards; turns out the `Object.keys(...).forEach(...` approach is more performant behind the scenes).